src/workerd/io/tracer.c++

```diff
@@ -117,6 +115,45 @@ WorkerTracer::WorkerTracer(kj::Rc<PipelineTracer> parentPipeline,
       parentPipeline(kj::mv(parentPipeline)),
       maybeTailStreamWriter(kj::mv(maybeTailStreamWriter)) {}
 
+kj::Maybe<kj::Own<tracing::TailStreamWriter>> PipelineTracer::getStageTailStreamWriter(
+    kj::TaskSet& waitUntilTasks) {
+  // Construct trace worker interfaces across the pipeline stack through a callback interface.
+  // If we are in the top level pipeline and the logging parameter is available, we could create
+  // the streaming tail worker interfaces directly instead of using a callback, but having a
+  // single unified code path is cleaner.
+  kj::Vector<kj::Own<WorkerInterface>> traceWorkers;
+  addTracers(traceWorkers);
+
+  if (traceWorkers.size() == 0) {
+    return kj::none;
+  }
+
+  // Initialize streaming tail worker.
+  // TODO(streaming-tail): memory management is off here – we need to add this to the pipeline
```

Comment:
This thing is not great – note that it is not newly added in this PR, but just ported from downstream. Last time I tested removing this it seemed to not break tests, I can try removing it in a future PR.